compiler/rustc_hir_analysis/src/check/intrinsicck.rs

```diff
@@ -299,6 +299,24 @@ impl<'a, 'tcx> InlineAsmCtxt<'a, 'tcx> {
             }
         }
 
+        if let ty::RawPtr(_, mutability) = *ty.kind() {
+            let is_mut = match mutability {
+                hir::Mutability::Not => false,
+                hir::Mutability::Mut => true,
+            };
+
+            if is_mut && asm.options.contains(InlineAsmOptions::READONLY) {
+                let msg = "Passing a mutable pointer to asm!() with 'readonly' option";
+                let note = "Mutable pointer suggest that this piece of assembly modifies the underlying object, consider using const pointer or checking the options";
+                self.tcx.dcx().struct_span_warn(expr.span, msg).with_note(note).emit();
+            }
+            if asm.options.contains(InlineAsmOptions::NOMEM) {
+                let msg = "Passing a pointer to asm!() with 'nomem' option";
+                let note = "Pointer suggest that this piece of assembly reads the underlying object, consider using usize or checking the options";
```

Comment:
why is a usize preferable?

Reply:
That was just what came to my mind when writing this at midnight. The warning and note could definitely be better, for example now I'm thinking of this:
```
warning: passing a pointer to asm! block with `nomem` option
 --> $PATH:$LINE
  |
  | core::arch::asm!("mov {p}, {p}", p = in(reg) p, options(nomem, nostack, preserves_flags));
  |                                  ^
  = note: `nomem` must be only used when no memory read or write happens inside the asm! block.
  = note: This is not limited to global variables, it also includes passed pointers.
  = note: If this is intentional, consider casting the pointer to an integer.
  = note: If not, change the `nomem` attribute to `readonly`
```
Warning is concise and note explains in more detail what is going on and how to fix the issue. Of course, depending on the pointer mutability and asm options, they will be a bit different.

Reply:
Hmm.

Reply:
I think this is probably underexplained and should be omitted:
```
  = note: If this is intentional, consider casting the pointer to an integer.
```

Assembly is `unsafe` and we should be careful to eschew recommendations that aren't clearly applicable, because people will blindly take recommendations without thinking hard (no matter how much we would like to believe otherwise).

Further, people who work with assembly often have the interesting belief that pointers and integers are the same thing, so I can imagine someone misunderstanding this and the next line as somehow equivalent:
```
  = note: If this is intentional, consider casting the pointer to an integer.
  = note: If not, change the `nomem` attribute to `readonly`
```

So I think the suggestions should be more conservative: "remove nomem, replace it with readonly, or allow this lint".

Reply:
It is probably fine if there's a good way to explain _why_ the cast to integer would be okay (so that people don't do it if it's not okay), but it seems sus, honestly.